File: apps/web/src/lib/portfolio/models.test.ts
import { describe, expect, it } from 'vitest';
import { isOpen, lotCostBasis, parsePosition, parsePositionLot } from './models';

describe('parsePosition', () => {
	// This is the row shape PostgREST actually returns for the query in
	// repository.ts: numeric columns as strings, the joined instrument embedded
	// under its table name.
	const row = {
		instrument_id: '00000000-0000-0000-0000-0000000000b1',
		quantity: '9.00000000',
		cost_basis: '1790.00000000',
		avg_cost: '198.88888889',
		realized_pl: '240.00000000',
		first_acquired_on: '2026-01-12',
		last_transaction_on: '2026-05-15',
		instruments: { symbol: 'AAPL', name: 'Apple Inc.' }
	};

	it('reads string-encoded numerics without casting', () => {
		const position = parsePosition(row);
		expect(position.quantity).toBe(9);
		expect(position.costBasis).toBe(1790);
		expect(position.avgCost).toBeCloseTo(198.88888889, 8);
		expect(position.realizedPl).toBe(240);
	});

	it('lifts the embedded instrument', () => {
		const position = parsePosition(row);
		expect(position.symbol).toBe('AAPL');
		expect(position.name).toBe('Apple Inc.');
	});

	it('survives a row with the instrument join missing', () => {
		const position = parsePosition({ instrument_id: 'x', quantity: 1, cost_basis: 10 });
		expect(position.symbol).toBe('—');
		expect(position.name).toBeNull();
		expect(position.avgCost).toBeNull();
		expect(position.realizedPl).toBe(0);
		expect(position.firstAcquiredOn).toBeNull();
	});

	// A holding whose cost basis is genuinely zero (an opening_balance standing
	// in for shares older than the export window) must not read as missing.
	it('keeps a real zero distinct from a missing value', () => {
		const position = parsePosition({ ...row, cost_basis: '0', avg_cost: null });
		expect(position.costBasis).toBe(0);
		expect(position.avgCost).toBeNull();
	});
});

describe('isOpen', () => {
	const base = parsePosition({ instrument_id: 'x', quantity: '0', cost_basis: '0' });

	it('treats a fully-sold holding as closed', () => {
		expect(isOpen(base)).toBe(false);
		expect(isOpen({ ...base, quantity: 0.0001 })).toBe(true);
	});
});

describe('parsePositionLot', () => {
	it('derives the lot cost basis from string-encoded numerics', () => {
		const lot = parsePositionLot({
			id: 'lot-1',
			acquired_on: '2026-02-18',
			quantity: '4.00000000',
			cost_per_share: '210.00000000'
		});

		expect(lot.quantity).toBe(4);
		expect(lot.costPerShare).toBe(210);
		expect(lotCostBasis(lot)).toBe(840);
		expect(lot.acquiredOn?.getDate()).toBe(18);
	});
});

File: apps/mobile/test/portfolio/models_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:project_stocks/journal/models.dart';
import 'package:project_stocks/portfolio/models.dart';

void main() {
  group('Position.fromJson', () {
    test('reads the embedded instrument', () {
      final position = Position.fromJson({
        'instrument_id': 'inst-1',
        'quantity': 9,
        'cost_basis': 1790,
        'avg_cost': 198.8889,
        'realized_pl': 240,
        'first_acquired_on': '2026-01-12',
        'last_transaction_on': '2026-05-04',
        'instruments': {'symbol': 'AAPL', 'name': 'Apple Inc.'},
      });

      expect(position.symbol, 'AAPL');
      expect(position.name, 'Apple Inc.');
      expect(position.quantity, 9);
      expect(position.isOpen, isTrue);
    });

    // Postgres numeric can arrive as a string; the whole row must still parse.
    test('accepts numerics encoded as strings', () {
      final position = Position.fromJson({
        'instrument_id': 'inst-1',
        'quantity': '9.00000000',
        'cost_basis': '1790.00000000',
        'avg_cost': '198.88888889',
        'realized_pl': '240.00000000',
        'instruments': {'symbol': 'AAPL'},
      });

      expect(position.quantity, 9);
      expect(position.costBasis, 1790);
      expect(position.avgCost, closeTo(198.8889, 0.0001));
    });

    test('survives a missing instrument join', () {
      final position = Position.fromJson({
        'instrument_id': 'inst-1',
        'quantity': 0,
        'cost_basis': 0,
        'realized_pl': 0,
      });

      expect(position.symbol, '—');
      expect(position.name, isNull);
      expect(position.avgCost, isNull);
    });

    test('a fully-sold holding is closed but keeps its realized P/L', () {
      final position = Position.fromJson({
        'instrument_id': 'inst-1',
        'quantity': 0,
        'cost_basis': 0,
        'realized_pl': '512.50',
        'instruments': {'symbol': 'TSLA'},
      });

      expect(position.isOpen, isFalse);
      expect(position.realizedPl, 512.5);
    });
  });

  test('PositionLot derives its own cost basis', () {
    final lot = PositionLot.fromJson({
      'id': 'lot-1',
      'acquired_on': '2026-02-18',
      'quantity': '5',
      'cost_per_share': '210',
    });

    expect(lot.costBasis, 1050);
  });

  group('Thesis', () {
    test('is live only until superseded', () {
      expect(
        Thesis.fromJson({
          'id': 't1',
          'rationale': 'x',
          'superseded_at': null,
        }).isLive,
        isTrue,
      );
      expect(
        Thesis.fromJson({
          'id': 't1',
          'rationale': 'x',
          'superseded_at': '2026-05-04T09:00:00Z',
        }).isLive,
        isFalse,
      );
    });
  });

  test('a note without an instrument is portfolio-wide', () {
    expect(Note.fromJson({'id': 'n1', 'body': 'x'}).isPortfolioWide, isTrue);
    expect(
      Note.fromJson({
        'id': 'n1',
        'body': 'x',
        'instrument_id': 'inst-1',
      }).isPortfolioWide,
      isFalse,
    );
  });
}
